pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,948 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bgp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"net"
+	"reflect"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	discoveryinformers "k8s.io/client-go/informers/discovery/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	discoverylisters "k8s.io/client-go/listers/discovery/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+	"k8s.io/utils/strings/slices"
+
+	"antrea.io/antrea/pkg/agent/bgp"
+	"antrea.io/antrea/pkg/agent/bgp/gobgp"
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/types"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/apis/crd/v1beta1"
+	crdinformersv1a1 "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha1"
+	crdinformersv1b1 "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1beta1"
+	crdlistersv1a1 "antrea.io/antrea/pkg/client/listers/crd/v1alpha1"
+	crdlistersv1b1 "antrea.io/antrea/pkg/client/listers/crd/v1beta1"
+	"antrea.io/antrea/pkg/util/env"
+)
+
+const (
+	controllerName = "BGPPolicyController"
+	// How long to wait before retrying the processing of a BGPPolicy change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Disable resyncing.
+	resyncPeriod time.Duration = 0
+)
+
+const defaultBGPListenPort int32 = 179
+
+const (
+	ipv4Suffix = "/32"
+	ipv6Suffix = "/128"
+)
+
+const dummyKey = "dummyKey"
+
+type bgpPolicyState struct {
+	// The local BGP server.
+	bgpServer bgp.Interface
+	// The port on which the local BGP server listens.
+	listenPort int32
+	// The AS number used by the local BGP server.
+	localASN int32
+	// The router ID used by the local BGP server.
+	routerID string
+	// routes stores all BGP routers advertised to BGP peers.
+	routes sets.Set[bgp.Route]
+	// peerConfigs is a map that stores configurations of BGP peers. The map keys are the concatenated strings of BGP
+	// peer IP address and ASN (e.g., "192.168.77.100-65000", "2001::1-65000").
+	peerConfigs map[string]bgp.PeerConfig
+}
+
+type Controller struct {
+	ctx context.Context
```

Comment:
this struct field would benefit from a comment IMO

Reply:
I may not get what you mean totally.  Could you give more information about this?

Reply:
I should use ctx in a wrong way. According to the golang document: `Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:`.

Reply:
Thanks for removing it and for referencing the Go documentation.
It indeed seemed like an unusual pattern to me, which is why I was requesting that a comment be added to explain why it was necessary. But removing the field altogether and passing the context as a function parameter is of course the right approach.